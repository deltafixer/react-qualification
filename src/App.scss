.app-container {
  display: flex;
  flex-direction: column;
  font-size: 18px;

  > * {
    margin-bottom: 12px;
  }

  .number-of-results {
    font-weight: bold;
  }
}
