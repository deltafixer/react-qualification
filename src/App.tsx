import React, { useEffect, useState } from "react";
import { IGitHubRepo } from "./models/IGitHubRepo";
import "./App.scss";

const gitHubUserURL = "https://api.github.com/users/facebook/repos";
const fetchOptions = {
  method: "GET",
  headers: {
    "Content-Type": "application/json",
  },
};

const debounce = (func: any, wait: number, isImmediate = false) => {
  let timeout: number | NodeJS.Timeout | null;
  return (...args: any) => {
    const later = () => {
      timeout = null;
      if (!isImmediate) {
        func(...args);
      }
    };
    const callNow = isImmediate && !timeout;
    clearTimeout(timeout as number);
    (timeout as NodeJS.Timeout) = setTimeout(later, wait);
    if (callNow) {
      func(...args);
    }
  };
};

const fetchMyRepos = async (): Promise<IGitHubRepo[]> => {
  const res = await fetch(gitHubUserURL, fetchOptions);
  return res.json();
};

function App() {
  const [repos, setRepos] = useState([] as IGitHubRepo[]);
  const [loading, setLoading] = useState(true);
  const [searchString, setSearchString] = useState("");
  let counter = 0;

  useEffect(() => {
    fetchMyRepos().then((fetchedRepos) => {
      setRepos(fetchedRepos);
      setLoading(false);
    });
  }, []);

  const searchInitiated = debounce((value: string) => {
    counter = 0;
    setSearchString(value);
  }, 400);

  return loading ? (
    <div className="loading">Loading repositories...</div>
  ) : (
    <div className="app-container">
      <input
        type="text"
        autoFocus
        placeholder="Type here to search for repos..."
        onChange={(e) => searchInitiated(e.target.value)}
      />
      <div className="repos-container">
        {repos.map((repo, index) => {
          if (repo.name.toLowerCase().includes(searchString.toLowerCase())) {
            counter += 1;
            return <div key={index}>{repo.name}</div>;
          }
          return null;
        })}
      </div>
      {/* kinda using the advantage of renderer, if this div is placed above, using counter like this won't be possible */}
      <div className="number-of-results">Number of results:&nbsp;{counter}</div>
    </div>
  );
}

export default App;
