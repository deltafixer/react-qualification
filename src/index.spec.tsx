import React from "react";
import { shallow, ShallowWrapper } from "enzyme";
import App from "./App";

let wrapper: ShallowWrapper;

describe("App component tests", () => {
  beforeAll(() => {
    wrapper = shallow(<App />);
  });

  it("renders correctly", () => {
    expect(wrapper.debug()).toMatchSnapshot();
  });
});
