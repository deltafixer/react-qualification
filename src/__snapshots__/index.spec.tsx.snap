// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component tests renders correctly 1`] = `"<Repos />"`;
